File: gausspr.js
/*
    GAUSSIAN PRIMES
    Copyright(C) 2018 Dario Alejandro Alpern

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Gaussian Primes is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Alpertron Calculators.  If not, see <http://www.gnu.org/licenses/>.
*/
/* global cantShowCanvas */
/* global commonGraphicEvents */
/* global copyStr */
/* global drawGraphic */
/* global get */
/* global getHeight */
/* global getWidth */
/* global initMenubarEvents */
/* global instantiate */
/* global isNotSpecialKey */
/* global moveGraphic */
/* global ptrToString */
/* global setNewDimensionsForCanvas */
/* global startLowLevelCode */
/* global startOffset */
// In order to reduce the number of files to read from Web server, this 
// Javascript file includes both the Javascript in the main thread and the 
// Javascript that drives WebAssembly on its own Web Worker.
const none = "none";
const block = "block";
const inline = "inline";
let fileContents = 0;
let center = null;
let buffer, env, asm;
let zoom, zoomDone, imgData;
let /** HTMLCanvasElement */ canvas;
let zoomin, zoomout, centerX, centerY;
let isMouseDown;
let currentX, currentY, pixels;
let HEAPU8;
let prevX1stTouch, prevY1stTouch;
let prevX2ndTouch, prevY2ndTouch;
let beforeMinus = "";
let asmGetInformation;
let asmDrawPartialGraphic;
let asmMoveGraphic;
let asmNbrChanged;
let bitsCanvas;
let information;
let animform;
let animate;
let stop;
let xincr;
let yincr;
let incrX;
let incrY;
let oldX, oldY;
let delay;
let doanimate;
let cancelanim;
let applet;
let interval;

//##  asmJS goes here (do not change symbols at the left).

function updateGraphic(dummy, nbr)
{
  if (nbr === 1)
  {
    copyStr(startOffset, centerX.value + "\x01" + centerY.value);
  }
  let width = getWidth();
  let height = getHeight();
  asmNbrChanged(startOffset, nbr, width, height);
  drawGraphic(canvas.getContext("2d"), 0, 0, width, height);
}

function showInfo(ptrInfo)
{
  let infoText = ptrToString(ptrInfo).split("^");
  if (infoText[1] !== centerX.value)
  {
    centerX.value = infoText[1];
  } 
  if (infoText[2] !== centerY.value)
  {
    centerY.value = infoText[2];
  } 
  information.innerHTML = infoText[0];
}

// Perl script will change the fields to match asm.js.
function startLowLevelCodeCallback(asm)
{
  asmGetInformation = asm.a;
  asmDrawPartialGraphic = asm.a;
  asmMoveGraphic = asm.a;
  asmNbrChanged = asm.a;
  let getPixels = asm.a;
  return getPixels;
}

function cantShowCanvas()
{
  return false;
}

function keydown(evt)
{
  let target = evt.target || evt.srcElement;
  let key = evt.key;
  if (isNotSpecialKey(evt))
  {
    if (!evt.ctrlKey && !evt.altKey && !evt.metaKey)
    {                                  // No modifier key pressed.
      if (key >= "0" && key <= "9")
      {                                // Digit key has been pressed.
        if (target.value.length >= 10 ||
            (target.value.charAt(0) !== "-" && target.value.length >= 9))
        {                              // Number is too large.
          evt.preventDefault();        // Do not propagate this key.
        }
      }
      else if (key === "-")
      {                                // Key minus has been pressed.
        if (target.value.indexOf("-") >= 0)
        {                              // There is already a minus sign.
          evt.preventDefault();        // Do not propagate this key.
        }
        else
        {
          beforeMinus = target.value;
        }
      }
      else 
      {                                // Not backspace, tab, right or left arrow, insert or delete key.
        evt.preventDefault();          // Do not propagate this key.  
      }
    }
  }
}
    
function animation()
{
  let currX = oldX - incrX * zoom;
  let currY = oldY + incrY * zoom;
  let diffX = Math.floor(currX) - Math.floor(oldX);
  let diffY = Math.floor(currY) - Math.floor(oldY);
  if (diffX !== 0 || diffY !== 0)
  {
    moveGraphic(diffX, diffY);
    showInfo(asmGetInformation(-1, -1));
  }
  oldX = currX;
  oldY = currY;
}

function startUp()
{  
  canvas = get("canvas");
  zoomin = get("zoomin");
  zoomout = get("zoomout");
  centerX = get("centerX");
  centerY = get("centerY");
  information = get("info");
  animate = get("animate");
  animform = get("animform");
  stop = get("stop");
  xincr = get("xincr");
  yincr = get("yincr");
  delay = get("delay");
  doanimate = get("doanimate");
  cancelanim = get("cancelanim");
  applet = get("applet");
  zoom = 8;
  zoomDone = 0;
  isMouseDown = false;
  let domRect = canvas.getBoundingClientRect();
  canvas.width = domRect.width;
  canvas.height = domRect.height;
  startLowLevelCode();
  commonGraphicEvents();
  centerX.onkeydown = function(evt)
  {
    if (evt.key === "Up" || evt.key === "ArrowUp")
    {
      canvas.focus();
      evt.preventDefault();          // Do not propagate this key.
    }
    keydown(evt);
  };
  centerX.oninput = function()
  {
    if (beforeMinus !== "" && centerX.value !== "-" + beforeMinus)
    {     // Minus sign is not in the first character. Move it to first character.
      centerX.value = "-" + beforeMinus;
    }
    beforeMinus = "";
    get("info").innerHTML = "";	
    updateGraphic(0, 1);
  };
  centerY.onkeydown = function(evt)
  {
    if (evt.key === "Up" || evt.key === "ArrowUp")
    {
      canvas.focus();
      evt.preventDefault();          // Do not propagate this key.
    }
    keydown(evt);
  };
  centerY.oninput = function()
  {
    if (beforeMinus !== "" && centerY.value !== "-" + beforeMinus)
    {     // Minus sign is not in the first character. Move it to first character.
      centerY.value = "-" + beforeMinus;
    }
    beforeMinus = "";
    get("info").innerHTML = "";
    updateGraphic(0, 1);
  };
  window.onresize = function()
  {
    setNewDimensionsForCanvas();
    updateGraphic(0, 1);
  };
  animate.onclick = function()
  {
    xincr.value = "0";
    yincr.value = "0";
    delay.value = "1";
    applet.style.display = none;
    animform.style.display = block;
    xincr.focus();
  };
  doanimate.onclick = function()
  {
    animform.style.display = none;
    applet.style.display = inline;
    animate.style.display = none;
    stop.style.display = inline;
    incrX = parseFloat(xincr.value);
    incrY = parseFloat(yincr.value);
    oldX = 0;
    oldY = 0;
    setNewDimensionsForCanvas();
    updateGraphic(0, 1);
    interval = setInterval(animation, parseFloat(delay.value) * 1000);    
  };
  cancelanim.onclick = function()
  {
    animform.style.display = none;
    applet.style.display = block;
  };
  stop.onclick = function()
  {
    stop.style.display = none;
    animate.style.display = inline;
    clearInterval(interval);
  };
  initMenubarEvents();
  if (get("left") == null)
  {
    setNewDimensionsForCanvas();
    updateGraphic(0, 1);
  }
  xincr.onkeydown = function(evt)
  {
    if (evt.key === "Down" || evt.key === "ArrowDown")
    {
      yincr.focus();
      evt.preventDefault();          // Do not propagate this key.
    }
  };
  yincr.onkeydown = function(evt)
  {
    if (evt.key === "Up" || evt.key === "ArrowUp")
    {
      xincr.focus();
      evt.preventDefault();          // Do not propagate this key.
    }
    if (evt.key === "Down" || evt.key === "ArrowDown")
    {
      delay.focus();
      evt.preventDefault();          // Do not propagate this key.
    }
  };
  delay.onkeydown = function(evt)
  {
    if (evt.key === "Up" || evt.key === "ArrowUp")
    {
      yincr.focus();
      evt.preventDefault();          // Do not propagate this key.
    }
    if (evt.key === "Down" || evt.key === "ArrowDown")
    {
      doanimate.focus();
      evt.preventDefault();          // Do not propagate this key.
    }
  };
}

window.addEventListener("load", startUp);

File: gausspr.c
//
// This file is part of Alpertron Calculators.
//
// Copyright 2015-2021 Dario Alejandro Alpern
//
// Alpertron Calculators is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Alpertron Calculators is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Alpertron Calculators.  If not, see <http://www.gnu.org/licenses/>.
//
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdint.h>
#include <stdbool.h>
#include "isprime.h"
#include "graphics.h"
#if !defined(__EMSCRIPTEN__) && TEST_GRAPHICS
  #include <SDL.h>
#endif

#define SMALL_NUMBER_BOUND 32768
#if !defined(__EMSCRIPTEN__) && TEST_GRAPHICS
  SDL_Surface* doubleBuffer;
  int oldXCenter;
  int oldYCenter;
  int oldXFraction;
  int oldYFraction;
#else
  #ifdef __ANDROID__
    unsigned int *pixelArrPtr;
  #else          // Emscripten
    unsigned int pixelArray[PIXEL_ARRAY_SIZE];
  #endif
#endif

#define MAX_LINES  1000
#define MAX_COLUMNS 2000
static char infoText[500];
extern int thickness;                 // Number of bits to shift.
extern int xCenter;
extern int xFraction;                 // Range of fraction: 0 to thickness - 1
extern int yCenter;
extern int yFraction;
extern int width;
extern int height;

 // A Gaussian number x+iy is prime if one of these 3 conditions is true:
 // 1) x is not zero and y is not zero and x^2+y^2 is prime
 // 2) x=0, y is prime and y=3 (mod 4)
 // 3) y=0, x is prime and x=3 (mod 4)
void setPointGaussian(unsigned int *ptrPixelArr, int x, int y)
{
  int xPhysical;
  int yPhysical;
  int row;
  int col;
  int firstRow;
  int firstCol;
  int lastRow;
  int lastCol;
  int firstRow2;
  int lastRow2;
  int firstCol2;
  int lastCol2;
  int xSquared[NBR_LIMBS];
  int ySquared[NBR_LIMBS];
  int check[NBR_LIMBS];
  unsigned int *ptrPixel;
  unsigned int color;
#if !defined(__EMSCRIPTEN__) && TEST_GRAPHICS
  Uint32 colorBlack = SDL_MapRGBA(doubleBuffer->format, 0, 0, 0, 255);
  Uint32 colorGreen = SDL_MapRGBA(doubleBuffer->format, 0, 192, 0, 255);
  Uint32 colorWhite = SDL_MapRGBA(doubleBuffer->format, 192, 192, 192, 255);
#else
  unsigned int colorBlack = 0xFF000000U;
  unsigned int colorGreen = 0xFF00C000U;
  unsigned int colorWhite = 0xFFC0C0C0U;
#endif
  xPhysical = (width / 2) + ((x - xCenter) * thickness) - xFraction;
  yPhysical = (height / 2) - ((y - yCenter) * thickness) + yFraction;
  color = colorBlack;          // Indicate not prime in advance.
  if (x == 0)
  {   // Number is imaginary.
    check[0] = ((y >= 0)? y : -y);
    check[1] = 0;
    if (((check[0] & 0x03) == 3) && isPrime(check))
    {                          // Number is gaussian prime.
      color = colorGreen;
    }
  }
  else if (y == 0)
  {   // Number is real.
    check[0] = ((x >= 0)? x : -x);
    check[1] = 0;
    if (((check[0] & 0x03) == 3) && isPrime(check))
    {                          // Number is gaussian prime.
      color = colorGreen;
    }
  }
  else
  {   // Number is not real or imaginary.
    multiplyBigNbrs(x, x, xSquared);
    multiplyBigNbrs(y, y, ySquared);
    AddBigNbrs(xSquared, ySquared, check);   // Check x^2 + y^2.
    if (isPrime(check))
    {                          // Number is gaussian prime.
      color = colorGreen;
    }
  }
  firstCol = ((xPhysical < 0)? 0: xPhysical);
  firstRow = ((yPhysical < 0)? 0: yPhysical);
  lastCol = xPhysical + thickness;
  if (lastCol > width)
  {
    lastCol = width;
  }
  lastRow = yPhysical + thickness;
  if (lastRow > height)
  {
    lastRow = height;
  }
  for (row = firstRow; row < lastRow; row++)
  {
    ptrPixel = ptrPixelArr + (row * ROW_WIDTH) + firstCol;
    for (col = firstCol; col < lastCol; col++)
    {
      *ptrPixel = color;
      ptrPixel++;
    }
  }
  if (thickness >= 4)
  {
    color = colorWhite;
    if (x == 0)
    {                     // Draw Y axis if possible.
      col = xPhysical + (thickness / 2);
      if ((col >= 0) && (col < width))
      {
        ptrPixel = ptrPixelArr + (firstRow * ROW_WIDTH) + col;
        for (row = firstRow; row < lastRow; row++)
        {
          *ptrPixel = color;
          ptrPixel += ROW_WIDTH;
        }
      }
    }
    else if ((x % 10) == 0)
    {
      col = xPhysical + (thickness / 2);
      if ((col >= 0) && (col < width))
      {
        firstRow2 = yPhysical + (thickness / 4);
        lastRow2 = firstRow2 + (thickness / 2);
        firstRow2 = ((firstRow2 < 0) ? 0 : firstRow2);
        if (lastRow2 > height)
        {
          lastRow2 = height;
        }
        ptrPixel = ptrPixelArr + (firstRow2 * ROW_WIDTH) + col;
        for (row = firstRow2; row < lastRow2; row++)
        {
          *ptrPixel = color;
          ptrPixel += ROW_WIDTH;
        }
      }
    }
    else
    {                     // Nothing to do.
    }
    if (y == 0)
    {                     // Draw X axis if possible.
      row = yPhysical + (thickness / 2);
      if ((row >= 0) && (row < height))
      {
        ptrPixel = ptrPixelArr + (row * ROW_WIDTH) + firstCol;
        for (col = firstCol; col < lastCol; col++)
        {
          *ptrPixel = color;
          ptrPixel++;
        }
      }
    }
    else if ((y % 10) == 0)
    {
      row = yPhysical + (thickness / 2);
      if ((row >= 0) && (row < height))
      {
        firstCol2 = xPhysical + (thickness / 4);
        lastCol2 = firstCol2 + (thickness / 2);
        firstCol2 = ((firstCol2 < 0)? 0 : firstCol2);
        if (lastCol2 > width)
        {
          lastCol2 = width;
        }
        ptrPixel = ptrPixelArr + (row * ROW_WIDTH) + firstCol2;
        for (col = firstCol2; col < lastCol2; col++)
        {
          *ptrPixel = color;
          ptrPixel++;
        }
      }
    }
    else
    {              // Nothing to do.
    }
  }
}     /* end method setPoint */

char *gaussianGetInformation(int x, int y)
{
  char *ptrText = infoText;
  
  infoText[0] = 0;   // Empty string.
  if (x >= 0)
  {
    int adjust = 16384 / thickness;
    int xLogical = xCenter + ((xFraction + x - (width / 2) + 16384) / thickness) - adjust;
    int yLogical = yCenter + 1 + ((yFraction - y + (height / 2) + 16384) / thickness) - adjust;
    ptrText = appendInt(infoText, xLogical);
    if (yLogical >= 0)
    {
      *ptrText = ' ';
      ptrText++;
      *ptrText = '+';
      ptrText++;
      *ptrText = ' ';
      ptrText++;
      ptrText = appendInt(ptrText, yLogical);
    }
    else
    {
      *ptrText = ' ';
      ptrText++;
      *ptrText = '-';
      ptrText++;
      *ptrText = ' ';
      ptrText++;
      ptrText = appendInt(ptrText, -yLogical);
    }
    *ptrText = 'i';
    ptrText++;
  }
  *ptrText = '^';       // Append separator.
  ptrText++;
  ptrText = appendInt(ptrText, xCenter);
  *ptrText = '^';       // Append separator.
  ptrText++;
  ptrText = appendInt(ptrText, yCenter);
  *ptrText = 0;
  return infoText;
}

static int getValue(const char **ppValue)
{
  const char *ptrValue = *ppValue;
  int nbr = 0;
  int sign = 0;
  if (*ptrValue == '-')
  {
    sign = 1;
    ptrValue++;
  }
  while (*ptrValue >= '0')
  {
    nbr = (nbr * 10) + (*ptrValue - '0');
    ptrValue++;
  }
  *ppValue = ptrValue;
  return sign? -nbr: nbr;
}

char *gaussianNbrChanged(const char *value, int inputBoxNbr, int newWidth, int newHeight)
{
  const char* ptrValue = value;
  width = newWidth;
  height = newHeight;
  if (inputBoxNbr == 1)
  {           
    xCenter = getValue(&ptrValue);  // Changing center X.
    xFraction = 0;
    ptrValue++;
    yCenter = getValue(&ptrValue);  // Changing center Y.
    yFraction = 0;
  }
  else if (inputBoxNbr == 3)
  {           // Zoom in
    if (thickness == 32)
    {
      return NULL;
    }
    thickness *= 2;
    xFraction <<= 1;
    yFraction <<= 1;
  }
  else
  {           // Zoom out
    if (thickness == 1)
    {
      return NULL;
    }
    thickness /= 2;
    xFraction >>= 1;
    yFraction >>= 1;
  }
  return NULL;
}

#ifdef __EMSCRIPTEN__
  setPointFunc setPoint = setPointGaussian;
  getInfoFunc getInfo = gaussianGetInformation;
  nbrChangedFunc nbrChgd = gaussianNbrChanged;
#endif
